extras/rxjava2/src/main/java/org/asynchttpclient/extras/rxjava2/maybe/AbstractMaybeAsyncHandlerBridge.java

```diff
@@ -0,0 +1,170 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.reactivex.MaybeEmitter;
+import io.reactivex.exceptions.CompositeException;
+import io.reactivex.exceptions.Exceptions;
+
+/**
+ * Abstract base class that bridges events between the {@code Maybe} reactive base type and {@code AsyncHandlers}.
+ *
+ * When an event is received, it's first checked if the Rx stream has been disposed asynchronously. If so, request
+ * processing is {@linkplain #disposed() aborted}, otherwise, the event is forwarded to the {@linkplain #delegate()
+ * wrapped handler}.
+ *
+ * When the request is {@link AsyncHandler#onCompleted() completed}, the result produced by the wrapped instance is
+ * forwarded to the {@code Maybe}: If the result is {@code null}, {@link MaybeEmitter#onComplete()} is invoked,
+ * {@link MaybeEmitter#onSuccess(T)} otherwise.
+ *
+ * Any errors during request processing are forwarded via {@link MaybeEmitter#onError(Throwable)}.
+ *
+ * @param <T>
+ *            the result type produced by the wrapped {@code AsyncHandler} and emitted via RxJava
+ */
+public abstract class AbstractMaybeAsyncHandlerBridge<T> implements AsyncHandler<Void> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractMaybeAsyncHandlerBridge.class);
+
+    /**
+     * The Rx callback object that receives downstream events and will be queried for its
+     * {@link MaybeEmitter#isDisposed() disposed state} when Async HTTP Client callbacks are invoked.
+     */
+    protected final MaybeEmitter<T> emitter;
+
+    /**
+     * Indicates if the delegate has already received a terminal event.
+     */
+    private final AtomicBoolean delegateTerminated = new AtomicBoolean();
+
+    protected AbstractMaybeAsyncHandlerBridge(MaybeEmitter<T> emitter) {
+        this.emitter = requireNonNull(emitter);
+    }
+
+    @Override
+    public final State onBodyPartReceived(HttpResponseBodyPart content) throws Exception {
+        return emitter.isDisposed() ? disposed() : delegate().onBodyPartReceived(content);
+    }
+
+    @Override
+    public final State onStatusReceived(HttpResponseStatus status) throws Exception {
+        return emitter.isDisposed() ? disposed() : delegate().onStatusReceived(status);
+    }
+
+    @Override
+    public final State onHeadersReceived(HttpResponseHeaders headers) throws Exception {
+        return emitter.isDisposed() ? disposed() : delegate().onHeadersReceived(headers);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p>
+     * The value returned by the wrapped {@code AsyncHandler} won't be returned by this method, but emtited via RxJava.
+     * </p>
+     *
+     * @return always {@code null}
+     */
+    @Override
+    public final Void onCompleted() {
+        if (delegateTerminated.getAndSet(true)) {
+            return null;
+        }
+
+        final T result;
+        try {
+            result = delegate().onCompleted();
+        } catch (final Throwable t) {
```

Comment:
That's an RxJava pattern found in various places there, e.g. in [one of the the map operator implementations](https://github.com/ReactiveX/RxJava/blob/v2.0.8/src/main/java/io/reactivex/internal/operators/single/SingleMap.java#L57). If it's some "fatal" exception from RxJava's point of view, it will be rethrown via [`io.reactivex.exceptions.Exceptions.throwIfFatal(…)`](https://github.com/ReactiveX/RxJava/blob/v2.0.8/src/main/java/io/reactivex/exceptions/Exceptions.java#L64) in `emitOnError`. Everything else will be reported via `onError`.